codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3UpdateEndpoint.java

```diff
@@ -0,0 +1,143 @@
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.ListUtils;
+
+public class S3UpdateEndpoint implements GoIntegration {
+    private static final String USE_PATH_STYLE_OPTION = "UsePathStyle";
+    private static final String UPDATE_ENDPOINT_ADDER = "addUpdateEndpointMiddleware";
+    private static final String UPDATE_ENDPOINT_INTERNAL_ADDER = "UpdateEndpoint";
+    private static final String GET_BUCKET_FROM_INPUT = "getBucketFromInput";
+
+    /**
+     * Gets the sort order of the customization from -128 to 127, with lowest
+     * executed first.
+     *
+     * @return Returns the sort order, defaults to -40.
+     */
+    @Override
+    public byte getOrder() {
+        return 127;
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ServiceShape service = settings.getService(model);
+        if (!isS3Service(model, service)) {
+            return;
+        }
+
+        goDelegator.useShapeWriter(service, this::writeMiddlewareHelper);
+
+        goDelegator.useShapeWriter(service, writer -> {
+            writeInputGetter(writer, model, symbolProvider, service);
+        });
+    }
+
+    private void writeInputGetter(GoWriter writer, Model model, SymbolProvider symbolProvider, ServiceShape service) {
+        writer.writeDocs("getBucketFromInput returns a boolean indicating if the input has a modeled bucket name, " +
+                " and a pointer to string denoting a provided bucket member value");
+        writer.openBlock("func getBucketFromInput(input interface{}) (*string, bool) {","}", ()-> {
+            writer.openBlock("switch i:= input.(type) {", "}", () -> {
+                service.getAllOperations().forEach((operationId)-> {
+                    OperationShape operation = model.expectShape(operationId, OperationShape.class);
+                    StructureShape input = model.expectShape(operation.getInput().get(), StructureShape.class);
+
+                    // get the member that targets "com.amazonaws.s3#BucketName" shape
+                    input.getAllMembers().values().stream().forEach((shape)->{
+                        if (shape.getTarget().getName().equals("BucketName")) {
+                            writer.write("case $P: return i.$L, true", symbolProvider.toSymbol(input), shape.getMemberName());
+                        }
+                    });
```

Comment:
We should probably ensure that only one member in the top level of the shape is of type BucketName and throw a codgen exception if there is two or more.

Reply:
I thought of this, but was not inclined to assert service models in the SDK gen. This should be handled by Smithy IMO. 

I will add the assertion here.